File: 70P/TwoSum.py
from typing import List 


# slower 
# time complexity (n * n = O(n ^ 2)) 
def twoSum(nums: List[int], target: int) -> List[int]:
    indexes = []
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                indexes.extend([i, j])
    return indexes
    




#faster O(N)

def two_sum(nums: List[int] = [2, 11, 7, 15], target: int = 9):
    hash_map = {}
    
    for i, v in enumerate(nums):
        diff = target - v 
        if diff in hash_map:
            return [i, hash_map[diff]]

        hash_map[v] = i


        
        
print(two_sum())

File: 70P/TwoSum.ts
let numbers1: number[] = [2, 11, 7, 15];
const tar: number = 9;
// O(N)
function twoSums(nums: number[], target: number): number[]{
    const hashMap: Record<number, number> = {}
    for (let i = 0; i < nums.length; i++){
        const diff: number = target - nums[i]
        if (diff in hashMap){
            return [i, hashMap[diff]];
        }
        hashMap[nums[i]] = i
    }
    return []
}

// O(n * n)
function twoSumBruteForce(nums: number[], target: number): number[]{
    const ans: number[] = [];
    for(let i = 0; i < nums.length; i++){
        for(let j = i + 1; j < nums.length; j++){
            if(nums[i] + nums[j] == target){
                ans.push(i, j);
            }
        }
    }
    return ans;
}

console.log(twoSums(numbers1, tar));
console.log(twoSumBruteForce(numbers1, tar));

